backends/avx/ceed-avx-basis.c

```diff
@@ -22,35 +22,149 @@
-                                   CeedInt C, CeedInt J,
-                                   const CeedScalar *restrict t,
-                                   CeedTransposeMode tmode, const CeedInt Add,
-                                   const CeedScalar *restrict u, CeedScalar *restrict v) {
+static int CeedTensorContract_Avx_Serial(Ceed ceed, CeedInt A, CeedInt B,
+                                         CeedInt C, CeedInt J,
+                                         const CeedScalar *restrict t,
+                                         CeedTransposeMode tmode,
+                                         const CeedInt Add,
+                                         const CeedScalar *restrict u,
+                                         CeedScalar *restrict v) {
   CeedInt tstride0 = B, tstride1 = 1;
   if (tmode == CEED_TRANSPOSE) {
     tstride0 = 1; tstride1 = J;
   }
 
+  const int JJ = 4, CC=8;
+
   if (!Add)
     for (CeedInt q=0; q<A*J*C; q++)
       v[q] = (CeedScalar) 0.0;
 
-  for (CeedInt a=0; a<A; a++)
-    for (CeedInt b=0; b<B; b++)
+  if (C != 1) {
+    // V = T U
+    for (CeedInt a=0; a<A; a++) {
+      // Blocks of 4
+      for (CeedInt j=0; j<(J/JJ)*JJ; j+=JJ) {
+        for (CeedInt c=0; c<(C/CC)*CC; c+=CC) {
+          __m256d vv[JJ][CC/4]; // Output tile to be held in registers
+          for (CeedInt jj=0; jj<JJ; jj++)
+            for (CeedInt cc=0; cc<CC/4; cc++)
+              vv[jj][cc] = _mm256_loadu_pd(&v[(a*J+j+jj)*C+c+cc*4]);
+
+          for (CeedInt b=0; b<B; b++) {
+            for (CeedInt jj=0; jj<JJ; jj++) { // unroll
+              __m256d tqv = _mm256_set1_pd(t[(j+jj)*tstride0 + b*tstride1]);
+              for (CeedInt cc=0; cc<CC/4; cc++) { // unroll
+                vv[jj][cc] += _mm256_mul_pd(tqv,
+                                _mm256_loadu_pd(&u[(a*B+b)*C+c+cc*4]));
+              }
+            }
+          }
+          for (CeedInt jj=0; jj<JJ; jj++)
+            for (CeedInt cc=0; cc<CC/4; cc++)
+              _mm256_storeu_pd(&v[(a*J+j+jj)*C+c+cc*4], vv[jj][cc]);
+        }
+      }
+      // JJ Remainder
+      CeedInt j=(J/JJ)*JJ;
+      for (CeedInt c=0; c<(C/CC)*CC; c+=CC) {
+        __m256d vv[JJ][CC/4]; // Output tile to be held in registers
+        for (CeedInt jj=0; jj<J-j; jj++)
+          for (CeedInt cc=0; cc<CC/4; cc++)
+            vv[jj][cc] = _mm256_loadu_pd(&v[(a*J+j+jj)*C+c+cc*4]);
+
+        for (CeedInt b=0; b<B; b++) {
+          for (CeedInt jj=0; jj<J-j; jj++) { // unroll
+            __m256d tqv = _mm256_set1_pd(t[(j+jj)*tstride0 + b*tstride1]);
+            for (CeedInt cc=0; cc<CC/4; cc++) { // unroll
+              vv[jj][cc] += _mm256_mul_pd(tqv,
+                              _mm256_loadu_pd(&u[(a*B+b)*C+c+cc*4]));
+            }
+          }
+        }
+        for (CeedInt jj=0; jj<J-j; jj++)
+          for (CeedInt cc=0; cc<CC/4; cc++)
+            _mm256_storeu_pd(&v[(a*J+j+jj)*C+c+cc*4], vv[jj][cc]);
+      }
+    }
+    for (CeedInt a=0; a<A; a++) {
```

Comment:
This will be used exclusively when 'C<8' so this is much better for high order elements.